File: run/bak/cpp/ffi/Test/RBTreeFFI.cc
#include "purescript.h"
#include <memory>
#include <algorithm>
namespace {
    volatile int dummy;
    struct Tree { bool black; std::shared_ptr<Tree> left; int val; std::shared_ptr<Tree> right; };
    auto red(std::shared_ptr<Tree> t) -> bool { return t && !t->black; }
    auto balance(bool b, std::shared_ptr<Tree> l, int v, std::shared_ptr<Tree> r) -> std::shared_ptr<Tree> {
        if (b) {
            if (red(l)) {
                if (red(l->left)) return std::make_shared<Tree>(Tree{false, std::make_shared<Tree>(Tree{true, l->left->left, l->left->val, l->left->right}), l->val, std::make_shared<Tree>(Tree{true, l->right, v, r})});
                if (red(l->right)) return std::make_shared<Tree>(Tree{false, std::make_shared<Tree>(Tree{true, l->left, l->val, l->right->left}), l->right->val, std::make_shared<Tree>(Tree{true, l->right->right, v, r})});
            }
            if (red(r)) {
                if (red(r->left)) return std::make_shared<Tree>(Tree{false, std::make_shared<Tree>(Tree{true, l, v, r->left->left}), r->left->val, std::make_shared<Tree>(Tree{true, r->left->right, r->val, r->right})});
                if (red(r->right)) return std::make_shared<Tree>(Tree{false, std::make_shared<Tree>(Tree{true, l, v, r->left}), r->val, std::make_shared<Tree>(Tree{true, r->right->left, r->right->val, r->right->right})});
            }
        }
        return std::make_shared<Tree>(Tree{b, l, v, r});
    }
    auto insRed(int v, std::shared_ptr<Tree> t) -> std::shared_ptr<Tree> {
        if(!t) return std::make_shared<Tree>(Tree{false, nullptr, v, nullptr});
        if(v < t->val) return balance(t->black, insRed(v, t->left), t->val, t->right);
        if(v > t->val) return balance(t->black, t->left, t->val, insRed(v, t->right));
        return std::make_shared<Tree>(Tree{t->black, t->left, t->val, t->right});
    }
    auto insert(int v, std::shared_ptr<Tree> t) -> std::shared_ptr<Tree> {
        auto res = insRed(v, t);
        return std::make_shared<Tree>(Tree{true, res->left, res->val, res->right});
    }
    auto depth(std::shared_ptr<Tree> t) -> int { return !t ? 0 : 1 + std::max(depth(t->left), depth(t->right)); }
}
FOREIGN_BEGIN( Test_RBTreeFFI )
exports["runRBTreeFFI"] = [](const boxed& in) -> boxed { 
    int n = unbox<int>(in); std::shared_ptr<Tree> t = nullptr;
    for(int i=n; i>0; i--) t = insert(i, t);
    int r = depth(t); dummy = r; return r; 
};
FOREIGN_END

File: scratch/records-tast-opportunity-20260916/full-scalar/src/Test/RBTreeFFI.java
private static final class Tree {
    final boolean black;
    final Tree left;
    final int value;
    final Tree right;

    Tree(boolean black, Tree left, int value, Tree right) {
        this.black = black;
        this.left = left;
        this.value = value;
        this.right = right;
    }
}

private static boolean red(Tree tree) {
    return tree != null && !tree.black;
}

private static Tree balance(boolean black, Tree a, int x, Tree b) {
    if (black) {
        if (red(a)) {
            if (red(a.left)) {
                Tree child = a.left;
                return new Tree(false,
                    new Tree(true, child.left, child.value, child.right), a.value,
                    new Tree(true, a.right, x, b));
            }
            if (red(a.right)) {
                Tree child = a.right;
                return new Tree(false,
                    new Tree(true, a.left, a.value, child.left), child.value,
                    new Tree(true, child.right, x, b));
            }
        }
        if (red(b)) {
            if (red(b.left)) {
                Tree child = b.left;
                return new Tree(false,
                    new Tree(true, a, x, child.left), child.value,
                    new Tree(true, child.right, b.value, b.right));
            }
            if (red(b.right)) {
                Tree child = b.right;
                return new Tree(false,
                    new Tree(true, a, x, b.left), b.value,
                    new Tree(true, child.left, child.value, child.right));
            }
        }
    }
    return new Tree(black, a, x, b);
}

private static Tree insertRed(int value, Tree tree) {
    if (tree == null) return new Tree(false, null, value, null);
    if (value < tree.value) {
        return balance(tree.black, insertRed(value, tree.left), tree.value, tree.right);
    }
    if (value > tree.value) {
        return balance(tree.black, tree.left, tree.value, insertRed(value, tree.right));
    }
    return new Tree(tree.black, tree.left, tree.value, tree.right);
}

private static Tree insert(int value, Tree tree) {
    Tree result = insertRed(value, tree);
    return new Tree(true, result.left, result.value, result.right);
}

private static int depth(Tree tree) {
    return tree == null ? 0 : 1 + Math.max(depth(tree.left), depth(tree.right));
}

public static final java.util.function.Function<Object, Object> runRBTreeFFI = Bench.nativeBenchmark(input -> {
    Tree tree = null;
    for (int value = ((Number) input).intValue(); value > 0; value--) {
        tree = insert(value, tree);
    }
    return depth(tree);
});
